pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -362,6 +367,97 @@ func applyCounterOrListPolicy(c *autoscalingv1.CounterPolicy, l *autoscalingv1.L
 	return 0, false, errors.Errorf("unable to apply ListPolicy %v", l)
 }
 
+func applySchedulePolicy(s *autoscalingv1.SchedulePolicy, f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister, nodeCounts map[string]gameservers.NodeCount, currentTime time.Time) (int32, bool, error) {
+	// Ensure the scheduled autoscaler feature gate is enabled
+	if !runtime.FeatureEnabled(runtime.FeatureScheduledAutoscaler) {
+		return 0, false, errors.Errorf("cannot apply SchedulePolicy unless feature flag %s is enabled", runtime.FeatureScheduledAutoscaler)
+	}
+
+	if isScheduleActive(s, currentTime) {
+		return computeDesiredFleetSize(s.Policy, f, gameServerLister, nodeCounts)
+	}
+
+	return f.Status.Replicas, false, errors.New("Inactive Schedule, Policy Not Applicable")
+}
+
+func applyChainPolicy(c autoscalingv1.ChainPolicy, f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister, nodeCounts map[string]gameservers.NodeCount, currentTime time.Time) (int32, bool, error) {
+	// Ensure the scheduled autoscaler feature gate is enabled
+	if !runtime.FeatureEnabled(runtime.FeatureScheduledAutoscaler) {
+		return 0, false, errors.Errorf("cannot apply ChainPolicy unless feature flag %s is enabled", runtime.FeatureScheduledAutoscaler)
+	}
+
+	replicas := f.Status.Replicas
+	var limited bool
+	var err error
+
+	// Loop over all entries in the chain
+	for _, entry := range c {
+		switch entry.Type {
+		case autoscalingv1.SchedulePolicyType:
+			replicas, limited, err = applySchedulePolicy(entry.Schedule, f, gameServerLister, nodeCounts, currentTime)
+			// If no error was returned from the schedule policy (schedule is active), then return the values given
+			if err == nil {
+				return replicas, limited, nil
+			}
```

Comment:
Should we log in case of an error ?

Reply:
In the case where there's an error here, it's either A. schedule is inactive or B. schedule contains a webhook policy that return an error. If there's only a schedule policy/schedule policy then the last schedule failed will be logged by the controller.

This leads me to ask if an inactive schedule should return an error or should we be checking if it's inactive here. I was trying to avoid calling `isScheduleActive` twice (one in chain policy to check if we should apply it, and another in the actual method itself in the case where scheduled is a policy by itself) by returning an error.

But to answer your question, yes I think it should be logged regardless. I can pass the controller's event logger through to the fleetautoscalers `computeDesiredFleetSize` method and record events within the `applyChainPolicy` and `applySchedulePolicy` methods.

How does that sound? @markmandel @zmerlynn

Reply:
I would shy away from logging "normal" events as error, especially since you'd be logging it every sync interval. IMO, logging here can be delegated to the individual `apply` functions, since they have better knowledge of what's abnormal - e.g. `applySchedulePolicy` can log if there's some hard error not related to the schedule simply being inactive, but `applyWebhookPolicy` wil probably always log the error, since we don't expect it.

FWIW, if these were functions in separate packages, I might answer differently, but it's not hard within the same package just to have logging consistently done at one level or the other.

Reply:
you resolved this, it looks like